### PR TITLE
Add suberror for network errors, Fixes AB#3064171

### PR DESCRIPTION
**Pull request by rpdome**
- Make UrlConnectionHttpClient throw a ClientException with IO_ERROR and a specific suberror (Based on where it fails).
- Move OAuth2SubError from ServiceException to BaseException (and rename it to a generic "subError")

image

This will allow us to separate network errors from other IO_ERROR

[AB#3064171](https://identitydivision.visualstudio.com/fac9d424-53d2-45c0-91b5-ef6ba7a6bf26/_workitems/edit/3064171)

### CODE REVIEW

File: common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java [hunk 1]
```diff
@@ -257,7 +262,26 @@ public static URL makeUrlSilent(String urlString) {
...
+            throw new ClientException(ClientException.MALFORMED_URL, e.getMessage(), e);
+        }
+    }
+
     public static String removeTrailingSlash(@NonNull final String urlString) {
         return urlString.replaceFirst("/*$", "");
     }
+
+
+    public static ClientException getNetworkLayerClientException(
```
Comment (mohitc1) [inline common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java]: >getNetworkLayerClientException [](http://example.com/codeflow?start=34&length=30)

Is this getting used? Looks same as ConnectionError.getClientException(cause)

**Review by mohitc1 — APPROVED** [review]
:shipit:

File: common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java [hunk 2]
```diff
@@ -257,7 +262,26 @@ public static URL makeUrlSilent(String urlString) {
...
+            final ConnectionError connectionSubError,
+            final Throwable cause) {
+        final ClientException e = new ClientException(IO_ERROR,
+                "An IO error occurred in the network layer: ${cause.message}",
+                cause);
+        e.setSubErrorCode(connectionSubError.name());
```
Comment (mohitc1) [inline common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java]: >connectionSubError.name() [](http://example.com/codeflow?start=26&length=25)

Also shouldn't it be using value here (getValue()) and not name()?

File: common4j/src/main/com/microsoft/identity/common/java/exception/ConnectionError.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * Returns true if the given [Throwable] is a connection error.
+     **/
+    fun compare(throwable: Throwable): Boolean {
+        if (throwable !is ClientException){
+            return false
+        }
+
+        return this.value == throwable.subErrorCode
+    }
+}
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/exception/ConnectionError.kt]: nit: new line ending

File: common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java [hunk 4]
```diff
@@ -257,6 +262,14 @@ public static URL makeUrlSilent(String urlString) {
         }
     }
 
+    public static URL makeUrl(String urlString) throws ClientException {
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java]: javadoc

Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/UrlUtil.java]:
```suggestion
    public static URL makeUrl(@NonNull final String urlString) throws ClientException {
```

**Review by shahzaibj — APPROVED** [review]

**Pull request merged**